### ISSUE TITLE
Errors from credential helpers are silently swallowed

### ISSUE THREAD

**Issue by AttilaTheFun** [bazelbuild/bazel#20146]
### Description of the bug:

I encountered this issue while trying to use the mapbox SDK which requires .netrc authentication to download:
https://docs.mapbox.com/ios/maps/guides/install/

I use the rules_swift_package_manager rule set as a bazel module dependency to produce a repository rule from the Package.swift manifest in their repository.

It was able to generate the targets, but when those targets attempted to download the prebuilt mapbox framework with repository_ctx.download(), I got a 401 unauthorized error.

I initially inquired about adding support for .netrc credentials to their repository rule but I was instructed to use a credential helper instead.

However, it appears these do not work with bazel modules as I have not been able to get either the repository rule or just the http_file to use the credential helper to authenticate my request.

I *was* able to download the file with my netrc file in my home directory, without the credential helper, but this approach did not work for the repository rule.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I created an example repository to reproduce the issue with fewer dependencies:
https://github.com/AttilaTheFun/credential_helper_bug

Just run:
```
bazel query "@mapbox//file"
```

And you will get the error:
```
logan@Logans-MBP credential_helper_bug % bazel query "@mapbox//file"
Starting local Bazel server and connecting to it...
INFO: Repository _main~mapbox_extension~mapbox instantiated at:
  callstack not available
Repository rule http_file defined at:
  /private/var/tmp/_bazel_logan/4f3056103a4a134a5a3e4ec875ef6723/external/bazel_tools/tools/build_defs/repo/http.bzl:466:28: in <toplevel>
WARNING: Download from https://api.mapbox.com/downloads/v2/mobile-maps-core/releases/ios/packages/10.16.1/MapboxCoreMaps.xcframework-dynamic.zip failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 401 Unauthorized
ERROR: An error occurred during the fetch of repository '_main~mapbox_extension~mapbox':
   Traceback (most recent call last):
        File "/private/var/tmp/_bazel_logan/4f3056103a4a134a5a3e4ec875ef6723/external/bazel_tools/tools/build_defs/repo/http.bzl", line 173, column 33, in _http_file_impl
                download_info = ctx.download(
Error in download: java.io.IOException: Error downloading [https://api.mapbox.com/downloads/v2/mobile-maps-core/releases/ios/packages/10.16.1/MapboxCoreMaps.xcframework-dynamic.zip] to /private/var/tmp/_bazel_logan/4f3056103a4a134a5a3e4ec875ef6723/external/_main~mapbox_extension~mapbox/file/downloaded: GET returned 401 Unauthorized
```

### Which operating system are you running Bazel on?

macOS Sonoma 14.1 (23B74)

### What is the output of `bazel info release`?

release 6.4.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

N/A

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
N/A
```

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

I don't know if credential helpers ever worked with Bazel modules.

### Have you found anything relevant by searching the web?

No

**Comment by Wyverald** [comment]
Your credential helper *is* called. Since it exits immediately with a non-zero exit code, we try to connect anyway due to the logic here: https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpConnectorMultiplexer.java;l=137-139;drc=19c73bcedb8e73c6b1854ba2789cbe79f2e626b4

Arguable whether this is a good idea, but at least we know your credential helper was invoked.

**Comment by Wyverald** [comment]
cc @tjgq

**Comment by Wyverald** [comment]
At the very least, we should provide some sort of log output to indicate the IOException.

**Comment by AttilaTheFun** [comment]
@Wyverald I added the exit(1) because I wanted to see Bazel fail with an error. 

The proposal states:
> If the credential helper exits with any exit code != 0, or if its output does not follow the protocol (e.g., is invalid JSON), Bazel will abort the operation with an appropriate FailureDetail indicating the reason (i.e., the build fails). A credential helper may provide a human-readable error message in its standard error that Bazel will print.

From: https://github.com/bazelbuild/proposals/blob/main/designs/2022-06-07-bazel-credential-helpers.md
Continuing to attempt to connect if there is an error seems to directly contradict this.

In any case, I figured out the issue with my credential helper. I thought the command was passed via standard input as well, but actually it's a positional command line argument. When I updated my python script to check:

```
    # Read the command:
    if len(sys.argv) < 2:
        raise ValueError("missing command")
    command = sys.argv[1]
    command = command.strip()
    if command != "get":
        raise ValueError("invalid command: " + command + "\nallowed commands: get")
```

Once I did this, the credential helper worked as expected. I pushed these changes to my example repository if this helps anyone.

Let's keep this issue open to track fixing the error handling, though, because this was very difficult to debug without it.

**Comment by brentleyjones** [comment]
@Wyverald f052e79453afeefc46474641b3010226a3778700 adds a warning, which is better than nothing, but the proposal says it should be an error. Is there a reason for this difference? It feels unsafe to make a network connection if the credential helper fails.

**Comment by Wyverald** [comment]
I'll leave @tjgq to comment on the proposal difference.

**Comment by tjgq** [comment]
Yes, I think it makes sense for a credential helper failure to be an error (although I don't think ignoring the error is a security issue).

**Comment by brentleyjones** [comment]
> although I don't think ignoring the error is a security issue

I think if you have a credential helper that is erroring out, and a connection is still made, one might accidentally upload blobs to a cache OR BES in anonymous mode, making things public that should be private. That is a security risk in my opinion.

**Comment by hochhaus** [comment]
> Yes, I think it makes sense for a credential helper failure to be an error

Thanks @tjgq. Could this issue be reopened until this is converted to an error (from the current warning)?

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
